proposals/NNNN-parameter-packs.md

```diff
@@ -0,0 +1,754 @@
+
+A same-shape requirement always relates two root type parameter packs. Member types always have the same shape as the root type parameter pack, so `shape(T.A) == shape(U.B)` reduces to `shape(T) == shape(U)`.
+
+**Inference:** Same-shape requirements are inferred in one of three ways:
+
+1. An abstract same-type requirement implies a same-shape requirement between two type parameter packs.
+
+2. A concrete same-type requirement implies a same-shape requirement between the type parameter packs on the left hand side and all type parameter packs captured by the concrete type on the right hand side.
+
+3. Finally, a same-shape requirement is inferred between each pair of type parameter packs captured by a pack expansion type appearing in certain positions.
+
+The following positions are subject to the same-shape requirement inference in Case 3:
+
+* all types appearing in the requirements of a trailing `where` clause of a generic function
+* the parameter types and return type of a generic function
+
+Recall that if the pattern of a pack expansion type contains more than one type parameter pack, all type parameter packs must be known to have the same shape, as outlined in the [Type substitution](#type-substitution) section. Same-shape requirement inference ensures that these invariants are satisfied when the pack expansion type occurs in one of the two above positions.
+
+If a pack expansion type appears in any other position, all type parameter packs captured by the pattern type must already be known to have the same shape, otherwise an error is diagnosed.
+
+For example, `zip` is a generic function, and the return type `(repeat (each T, each U))` is a pack expansion type, therefore the same-shape requirement `shape(T) == shape(U)` is automatically inferred:
+
+```swift
+// Return type infers 'where length(T) == length(U)'
+func zip<each T, each U>(firsts: repeat each T, seconds: repeat each U) -> (repeat (each T, each U)) {
+  return (repeat (each firsts, each seconds))
+}
+
+zip(firsts: 1, 2, seconds: "hi", "bye") // okay
+zip(firsts: 1, 2, seconds: "hi") // error; length requirement is unsatisfied
+```
+
+Here is an example where the same-shape requirement is not inferred:
+
+```swift
+func foo<each T, each U>(t: repeat each T, u: repeat each U) {
+  let tup: (repeat (each T, each U)) = /* whatever */
+}
+```
+
+The type annotation of `tup` contains a pack expansion type `repeat (each T, each U)`, which is malformed because the requirement `count(T) == count(U)` is unsatisfied. This pack expansion type is not subject to requirement inference because it does not occur in one of the above positions.
+
+#### Open questions
+
+While type packs cannot be written directly, a requirement where both sides are concrete types is desugared using the type matching algorithm, therefore it will be possible to write down a requirement that constraints a type parameter pack to a concrete type pack, unless some kind of restriction is imposed:
```

Comment:
```suggestion
While type packs cannot be written directly, a requirement where both sides are concrete types is desugared using the type matching algorithm, therefore it will be possible to write down a requirement that constrains a type parameter pack to a concrete type pack, unless some kind of restriction is imposed:
```